File: Codeforces/div811/D2.py

import sys


for _ in range(int(input())):
    t = input()
    n = int(input())
    ss = [input() for __ in range(n)]
    s = set()
    for i in range(len(t)):
        for j in range(n):
            if t[i:].startswith(ss[j]):
                s.add((i, i + len(ss[j]), j))
    s = sorted(s, key=lambda x: (x[0], x[1]))
    # print(len(s))
    end = 0
    i = 0
    ans = []
    while True:
        if end >= len(t):
            break
        if i >= len(s) or s[i][0] > end:
            ans = -1
            break
        for ii in range(i + 1, len(s)):
            if s[ii][0] <= end:
                i = ii
        ans.append((s[i][2] + 1, s[i][0] + 1))
        end = s[i][1]

    if ans == -1:
        print(ans)
    else:
        print(len(ans))
        for i in ans:
            print(*i)

File: Codeforces/div811/D2.cpp

#include <iostream>
#include <vector>
#include <string>
#include <set>
#include <algorithm>

using namespace::std;

int main()
{
	int q;
	cin >> q;
	while (q--)
	{
		string t;
		cin >> t;
		int n;
		cin >> n;
		vector<string> ss(n);
		for (int i=0; i < n; ++i)
			cin >> ss.at(i);
		set<vector<int>> st;
		for (int i=0; i < t.size(); ++i)
		{
			for (int j=0; j < n; ++j)
			{
				if (t.substr(i, 1000).starts_with(ss.at(j)))
				{
					vector<int> ins = {i, static_cast<int>(i + ss.at(j).size()), j};
					st.insert(ins);
				}
			}
		}
		vector<vector<int>> s(st.begin(), st.end());
		sort(s.begin(), s.end(), [](
					const vector<int> &a, const vector<int> &b)
				{
					vector<int> aa = {a.at(0), a.at(1)};
					vector<int> bb = {b.at(0), b.at(1)};
					return aa < bb;
					});
		int end = 0;
		int i = 0;
		bool ok = true;
		vector<vector<int>> ans;
		while (true)
		{
			if (end >= t.size())
				break ;
			if (i >= s.size() or s.at(i).at(0) > end)
			{
				ok = false;
				break ;
			}
			for (int ii=i + 1; ii < s.size(); ++ii)
			{
				if (s.at(ii).at(0) <= end)
					i = ii;
			}
			vector<int> ins = {s.at(i).at(2) + 1, s.at(i).at(0) + 1};
			ans.push_back(ins);
			end = s.at(i).at(1);
		}

		if (not ok)
			cout << "-1\n";
		else
		{
			cout << ans.size() << endl;
			for (const auto &a : ans)
			{
				cout << a.at(0) << ' ' << a.at(1) << endl;
			}
		}
	}
}

